### PR TITLE
[introspection] Migrate .NET code to use the new .NET-style availability attributes.

### PR DESCRIPTION
**Pull request by rolfbjarne**
* This also means using the 'ApplePlatform' enum instead of the 'PlatformName'
  enum, because the latter will be removed in .NET.
* A few APIs and availability attributes had to be adjusted, since the updated
  tests catches more errors.

### CODE REVIEW

File: tests/common/PlatformInfo.cs [hunk 1]
```diff
@@ -91,92 +119,75 @@ static PlatformInfo GetHostPlatformInfo ()
...
+
+			foreach (var attr in attributes) {
+				if (!attr.TryParse (out ApplePlatform? platform, out var version))
+					continue;
+
+				if (targetPlatform.Name == ApplePlatform.MacCatalyst) {
+					if (platform != ApplePlatform.iOS && platform != ApplePlatform.MacCatalyst)
+						continue;
+				} else if (platform != targetPlatform.Name) {
+					continue;
```
Comment (chamons) [inline tests/common/PlatformInfo.cs:170]: Nit - Space inconsistency here.

**Review by chamons — APPROVED** [review]

File: tests/common/PlatformInfo.cs [hunk 2]
```diff
@@ -50,6 +52,20 @@ static PlatformInfo GetHostPlatformInfo ()
...
+			else if (name.StartsWith ("mac", StringComparison.Ordinal))
+				platformInfo.Name = ApplePlatform.MacOSX;
+			else if (name.StartsWith ("ios", StringComparison.Ordinal) || name.StartsWith ("iphoneos", StringComparison.Ordinal) || name.StartsWith ("ipados", StringComparison.Ordinal))
+				platformInfo.Name = ApplePlatform.iOS;
+			else if (name.StartsWith ("tvos", StringComparison.Ordinal))
+				platformInfo.Name = ApplePlatform.TVOS;
+			else if (name.StartsWith ("watchos", StringComparison.Ordinal))
+				platformInfo.Name = ApplePlatform.WatchOS;
+			else
+				throw new FormatException ($"Unknown product name: {name}");
```
Comment (mandel-macaque) [inline tests/common/PlatformInfo.cs:67]: does not matter, but you could have used a switch + pattern matching:
```csharp
switch(n)
{
    case string name when name.StartsWith("tvos", StringComparison.Ordinal)):
       return "123";
    case string name when name.StartsWith("watchos", StringComparison.Ordinal):
       return "123";
    default:
       return "ABCXYZ123";
}
```

Comment (Therzok) [reply tests/common/PlatformInfo.cs:67]: heh, might've been a way to reduce code duplication by having something like:

```csharp
#if NET
using PlatformName = ApplePlatform;
#endif
```

File: tools/common/OSPlatformAttributeExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+using System.Runtime.Versioning;
+
+using Xamarin.Utils;
+
+#nullable enable
+
+namespace Xamarin.Utils {
+	internal static class OSPlatformAttributeExtensions {
+#if NET
+		public static bool TryParse (string? platformName, [NotNullWhen (true)] out string? platform, out Version? version)
```
Comment (Therzok) [inline tools/common/OSPlatformAttributeExtensions.cs:12]: This attribute can be stubbed out with an internal version of the nullability attributes in the assembly:
https://source.dot.net/#Microsoft.Extensions.Logging.Console/NullableAttributes.cs,c0596421bc26d09e

Comment (Therzok) [reply tools/common/OSPlatformAttributeExtensions.cs:12]: It also works for codeanalysis as a result when targeting .netframework

File: tests/common/PlatformInfo.cs [hunk 4]
```diff
@@ -91,92 +119,75 @@ static PlatformInfo GetHostPlatformInfo ()
...
-				}
-				// FIXME - temporary, while older attributes co-exists (in manual bindings)
-				if (ca is AvailabilityBaseAttribute old)
-					list.Add (old);
-				if (ca is ObsoleteAttribute)
-					return false;
-			}
-			return list.IsAvailable (targetPlatform);
+			customAttributes = customAttributes.ToArray (); // don't iterate twice
+			if (customAttributes.Any (v => v is ObsoleteAttribute))
```
Comment (Therzok) [inline tests/common/PlatformInfo.cs:133]: LINQ here is making the implementation more complex than a normal foreach loop :D

File: tests/common/PlatformInfo.cs [hunk 5]
```diff
@@ -91,92 +119,75 @@ static PlatformInfo GetHostPlatformInfo ()
...
 
 		public static bool IsAvailable (this ICustomAttributeProvider attributeProvider, PlatformInfo targetPlatform)
 		{
+			var customAttributes = attributeProvider.GetCustomAttributes (true);
+
 #if NET
-			var list = new List<AvailabilityBaseAttribute> ();
-			foreach (var ca in attributeProvider.GetCustomAttributes (true)) {
-				if (ca is OSPlatformAttribute aa) {
-					var converted = aa.Convert ();
```
Comment (Therzok) [inline tests/common/PlatformInfo.cs:156]: Is conversion no longer needed?

**Review by dalexsoto — APPROVED** [review]

**Review by tj-devel709 — APPROVED** [review]

**Comment by rolfbjarne** [comment]
Test failure is unrelated (https://github.com/xamarin/maccore/issues/2414).

**Pull request merged**